File: 0077_combine.py
class Solution:
    def combine(self, n, k):
        result = []
        self.backtracking(n, k, 1, [], result)
        return result
    def backtracking(self, n, k, startindex, path, result):
        if len(path) == k:
            result.append(path[:])
            return
        for i in range(startindex, n - (k - len(path)) + 2):
            path.append(i)
            self.backtracking(n, k, i + 1, path, result)
            path.pop()

if __name__ == '__main__':
    n = int(input())
    k = int(input())
    print(Solution().combine(n, k))

File: problem/0077_combine.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
private:
    vector<vector<int>> result;
    vector<int> path;
    void backtracking(int n, int k, int startindex) {
        if (path.size() == k) {
            result.push_back(path);
            return;
        }
        for (int i = startindex; i <= n; i++) {
        // for (int i = startindex; i <= n - (k - path.size()) + 1; i++){    //优化该代码,优化的地方
            path.push_back(i);
            backtracking(n, k, i + 1);
            path.pop_back();
        }
    }
public:
    vector<vector<int>> combine(int n, int k){
        backtracking(n, k, 1);
        return result;
    }
};


int main(){
    Solution solution;
    int n;
    int k;
    cin >> n;
    cin >> k;
    vector<vector<int>> result = solution.combine(n, k);
    for (const auto& row : result) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }

}
